src/terminal/adapter/SixelParser.cpp

```diff
@@ -0,0 +1,840 @@
+{
+    return til::at(_colorTable, tableIndex);
+}
+
+constexpr RGBQUAD SixelParser::_makeRGBQUAD(const COLORREF color) noexcept
+{
+    return RGBQUAD{
+        .rgbBlue = GetBValue(color),
+        .rgbGreen = GetGValue(color),
+        .rgbRed = GetRValue(color),
+        .rgbReserved = 255
+    };
+}
+
+void SixelParser::_updateTextColors()
+{
+    // On the original hardware terminals, text and images shared the same
+    // color table, so palette changes made in an image would be reflected in
+    // the text output as well.
+    if (_conformanceLevel <= 3 && _maxColors > 2 && _colorTableChanged) [[unlikely]]
+    {
+        for (IndexType tableIndex = 0; tableIndex < _maxColors; tableIndex++)
+        {
+            _dispatcher.SetColorTableEntry(tableIndex, _colorFromIndex(tableIndex));
+        }
+        _colorTableChanged = false;
+    }
+}
+
+void SixelParser::_initImageBuffer()
+{
+    _imageBuffer.clear();
+    _imageOriginCell = _textCursor;
+    _imageCursor = {};
+    _imageWidth = 0;
+    _imageMaxWidth = _availablePixelWidth;
+    _imageLineCount = 0;
+    _resizeImageBuffer(_sixelHeight);
+
+    _lastFlushLine = 0;
+    _lastFlushTime = steady_clock::now();
+
+    // Prior to the VT340, the background was filled as soon as the sixel
+    // definition was started, because the initial raster attributes could
+    // not be altered.
+    if (_conformanceLevel < 3)
+    {
+        _fillImageBackground();
+    }
+}
+
+void SixelParser::_resizeImageBuffer(const til::CoordType requiredHeight)
+{
+    const auto requiredSize = (_imageCursor.y + requiredHeight) * _imageMaxWidth;
+    if (static_cast<size_t>(requiredSize) > _imageBuffer.size())
+    {
+        static constexpr auto transparentPixel = IndexedPixel{ .transparent = true };
+        _imageBuffer.resize(requiredSize, transparentPixel);
+    }
+}
+
+void SixelParser::_fillImageBackground()
+{
+    if (_backgroundFillRequired) [[unlikely]]
+    {
+        _backgroundFillRequired = false;
+
+        const auto backgroundHeight = std::min(_backgroundSize.height, _availablePixelHeight);
+        const auto backgroundWidth = std::min(_backgroundSize.width, _availablePixelWidth);
+        _resizeImageBuffer(backgroundHeight);
+
+        // When a background fill is requested, we prefill the buffer with the 0
+        // color index, up to the boundaries set by the raster attributes (or if
+        // none were given, up to the page boundaries). The actual image output
+        // isn't limited by the background dimensions though.
+        static constexpr auto backgroundPixel = IndexedPixel{};
+        const auto backgroundOffset = _imageCursor.y * _imageMaxWidth;
+        auto dst = std::next(_imageBuffer.begin(), backgroundOffset);
+        for (auto i = 0; i < backgroundHeight; i++)
+        {
+            std::fill_n(dst, backgroundWidth, backgroundPixel);
+            std::advance(dst, _imageMaxWidth);
+        }

```

Comment:
nit: I think this code and some other in ImageSlice could be simplified a bit if we had a stride-aware `memset` function.

Reply:
Maybe... but I don't think it's really worth it. It's not a massive chunk of code, and replacing a two line loop with a call to something like `stride_aware_memset` doesn't seem likely to make it more readable in my opinion. You can always change it when you're refactoring if you still think it's a good idea.